src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -121,49 +125,46 @@ private void ConfigureApplication(WebHostBuilderContext context, IApplicationBui
+                var targetRouteBuilder = GetEndpointRouteBuilder(_builtApplication);
 
-                    foreach (var ds in _builtApplication.DataSources)
-                    {
-                        routes.DataSources.Add(ds);
-                    }
-
-                    // Chain the execution of the source pipeline into the destination pipeline
-                    app.Use(next =>
-                    {
-                        _builtApplication.Run(next);
-                        return _builtApplication.BuildRequestDelegate();
-                    });
+                if (targetRouteBuilder is null)
+                {
+                    // The app defined endpoints without calling UseRouting() explicitly, so call UseRouting() implicitly.
+                    app.UseRouting();
 
-                    // Add a UseEndpoints at the end
-                    app.UseEndpoints(e => { });
+                    // An implicitly created IEndpointRouteBuilder was addeded to app.Properties by the UseRouting() call above.
+                    targetRouteBuilder = GetEndpointRouteBuilder(app)!;
+                    implicitRouting = true;
                 }
                 else
                 {
-                    // Since we register routes into the source pipeline's route builder directly,
-                    // if the user called UseRouting, we need to copy the data sources
-                    foreach (var ds in _builtApplication.DataSources)
-                    {
-                        _builtApplication.RouteBuilder.DataSources.Add(ds);
-                    }
-
-                    // We then implicitly call UseEndpoints at the end of the pipeline
+                    // UseRouting() was called explicitely, but we may still need to call UseEndpoints() implicitely at
+                    // the end of the pipeline.
                     _builtApplication.UseEndpoints(_ => { });
+                }
 
-                    // Wire the source pipeline to run in the destination pipeline
-                    app.Run(_builtApplication.BuildRequestDelegate());
+                // Copy the endpoints to the explicitly or implicitly created IEndopintRouteBuilder.
+                foreach (var ds in _builtApplication.DataSources)
+                {
+                    targetRouteBuilder.DataSources.Add(ds);
                 }
             }
-            else
+
+            // Wire the source pipeline to run in the destination pipeline
+            app.Use(next =>
+            {
+                _builtApplication.Run(next);
+                return _builtApplication.BuildRequestDelegate();
+            });
+
+            // Implicitly call UseEndpoints() at the end of the pipeline if UseRouting() was called implicitly.
+            // We could add this to the end of _buildApplication instead if UseEndpoints() was not so picky about
+            // being called with the same IApplicationBluilder instance as UseRouting().
```

Comment:
```suggestion
            // being called with the same IApplicationBulilder instance as UseRouting().
```

Reply:
This is probably more correct anyways. Is there any observable difference with this ordering?

Reply:
Nope. I just don't like that it's more lines of code 😄